File: 백준/Gold/32405. 배틀 로얄/배틀 로얄.py
import sys

input = sys.stdin.readline

# 입력 받기
n = int(input())
atk_list = list(map(int, input().split()))
hp_list = list(map(int, input().split()))

# 초기화
players = [(i + 1, atk, hp) for i, (atk, hp) in enumerate(zip(atk_list, hp_list))]
atk_sum = 0  # 공격력 누적합
loop = 0  # 라운드 수

while len(players) > 1:
    players = [player for player in players if player[2] > atk_sum - player[1] * loop and (atk_sum := atk_sum + player[1])]
    loop += 1

# 마지막 남은 플레이어 출력
print(players[0][0])

File: 백준/Gold/32405. 배틀 로얄/배틀 로얄.kt
package org.example

import java.util.StringTokenizer

data class Player(val id: Int, val atk: Int, val hp: Int)

fun main() {
    val br = System.`in`.bufferedReader()
    val N = br.readLine().toInt()
    val stD = StringTokenizer(br.readLine())
    val stH = StringTokenizer(br.readLine())

    var players = mutableListOf<Player>()

    for (i in 1..N) {
        players.add(Player(i, stD.nextToken().toInt(), stH.nextToken().toInt()))
    }
    //
    var atkSum = 0
    var loop = 0

    while (players.size > 1) {
        players = players.filter { player ->
            if (player.hp > atkSum - player.atk * loop) {
                atkSum += player.atk
                true
            } else false
        }.toMutableList()
        loop++
    }

    println(players[0].id)
    br.close()
}


